groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -704,55 +1010,415 @@ struct Optional_Data<TYPE, true> : public Optional_DataImp<TYPE> {
+    // 'optional' type.  The main template is instantiated for allocator-aware
+    // types and holds an allocator used to create all objects of 'value_type'
+    // managed by the 'optional' object.
 
-// MANIPULATORS
-#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
-template <class TYPE>
-template <class... ARGS>
-inline
-void Optional_DataImp<TYPE>::emplace(
-                         bslma::Allocator                           *allocator,
-                         BSLS_COMPILERFEATURES_FORWARD_REF(ARGS)...  args)
-{
-    reset();
-    BloombergLP::bslma::ConstructionUtil::construct(
-        d_buffer.address(),
-        allocator,
-        BSLS_COMPILERFEATURES_FORWARD(ARGS, args)...);
-    d_hasValue = true;
-}
+  public:
+    // PUBLIC TYPES
+    typedef TYPE value_type;
+        // 'value_type' is an alias for the underlying 'TYPE' upon which this
+        // template class is instantiated, and represents the type of the
+        // managed object.  The name is chosen so it is compatible with the
+        // 'std::optional' implementation.
+
+    typedef typename bsl::allocator<char> allocator_type;
+
+  private:
+    // PRIVATE TYPES
+    typedef BloombergLP::bslmf::MovableRefUtil MoveUtil;
+
+#ifndef BSLS_COMPILERFEATURES_SUPPORT_OPERATOR_EXPLICIT
+    // UNSPECIFIED BOOL
+
+    // This type is needed only in C++03 mode, where 'explicit' conversion
+    // operators are not supported.  A 'function' is implicitly converted to
+    // 'UnspecifiedBool' when used in 'if' statements, but is not implicitly
+    // convertible to 'bool'.
+    typedef BloombergLP::bsls::UnspecifiedBool<optional> UnspecifiedBoolUtil;
+    typedef typename UnspecifiedBoolUtil::BoolType       UnspecifiedBool;
+
+#endif
+
+    // DATA
+    BloombergLP::bslstl::Optional_Data<TYPE> d_value;
+        // in-place 'TYPE' object
+
+    allocator_type                           d_allocator;
+        // allocator to be used for all in-place 'TYPE' objects
+
+  public:
+    // TRAITS
+    BSLMF_NESTED_TRAIT_DECLARATION(optional,
+                                   BloombergLP::bslma::UsesBslmaAllocator);
+    BSLMF_NESTED_TRAIT_DECLARATION(optional,
+                                   BloombergLP::bslmf::UsesAllocatorArgT);
+    BSLMF_NESTED_TRAIT_DECLARATION_IF(
+        optional,
+        BloombergLP::bslmf::IsBitwiseMoveable,
+        BloombergLP::bslmf::IsBitwiseMoveable<TYPE>::value);
+
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+    BSLMF_NESTED_TRAIT_DECLARATION_IF(optional,
+                                      bsl::is_trivially_copyable,
+                                      bsl::is_trivially_copyable<TYPE>::value);
+    // Workaround for C++03 'bsl::is_trivially_copyable' trait.
+#endif  //BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+
+    // CREATORS
+    optional();
+        // Create a disengaged 'optional' object.  Use the currently installed
+        // default allocator to supply memory for future 'value_type' objects.
```

Comment:
```suggestion
        // default allocator to supply memory.
```
I think it's not necessary to talk about "future objects" since scoped allocator model implies it.  Please apply this change to other places as well.

Overall, I suggest copying the doc from nullablevalue with slight wording adjustments (e.g., using the "disengaged terminology ).

Reply:
doc from nullablevalue copied and adjusted for all of the constructors